Add "verify manifest" activity, fixes #35

djjuhasz opened the pull request.

jraddaoui commented on internal/activities/verify_manifest_test.go:
```diff
@@ -0,0 +1,136 @@
+	"github.com/artefactual-sdps/preprocessing-sfa/internal/sip"
+)
+
+func setup(t *testing.T) string {
+	t.Helper()
+
+	td := t.TempDir()
+
+	missingFile := filepath.Join(td, "missing_file")
+	err := cp.Copy("../testdata/little_digitized_AIP", missingFile)
+	if err != nil {
+		t.Fatalf("setup: copy: %v", err)
+	}
+
+	err = os.Remove(
+		filepath.Join(missingFile, "content", "content", "d_0000001", "00000001.jp2"),
+	)
+	if err != nil {
+		t.Fatalf("setup: remove: %v", err)
+	}
+
+	extraFile := filepath.Join(td, "extra_file")
+	err = cp.Copy("../testdata/little_digitized_AIP", extraFile)
+	if err != nil {
+		t.Fatalf("setup: copy: %v", err)
+	}
+
+	err = os.WriteFile(
+		filepath.Join(extraFile, "content", "content", "d_0000001", "extra_file.txt"),
+		[]byte("I'm an extra file."),
+		0o600,
+	)
+	if err != nil {
+		t.Fatalf("setup: write file: %v", err)
+	}
+
+	return td
+}
```
It would be nice to fake the SIP structure and XML contents instead of having one in that testdata folder, it shouldn't be hard to do, there are some examples across the tests already:

https://github.com/artefactual-sdps/preprocessing-sfa/blob/main/internal/activities/transform_sip_test.go

I doesn't need to be a full SIP with the XSD files and all to prove the activity behavior.

djjuhasz replied:
Yeah, good idea @jraddaoui.  I was originally going to check the file checksums too, which would require actual file contents to checksum, but after rereading the comments on #22 I noticed that we may be checking the metadata.xml checksums against the BagIt bag checksums after we generate the bag.  In any case, I think it's better to check the checksums in a separate activity.

TIL; I'll remove the test SIP and fake the XML + structure in the test file.

jraddaoui reviewed (approved):
Nice! Feel free to ignore the comment below.

djjuhasz commented:
@jraddaoui after removing the test SIP and rebasing to get the latest updates from main, I've made quite a few change to the original PR.  Can you take another look at the PR please?

jraddaoui commented on internal/sip/sip.go:
```diff
@@ -11,29 +11,49 @@ import (
-	TopLevelPaths       []string
+
+	// XSDPath is the filepath of the "xsd" directory.
+	XSDPath string
```
It's the path of the "arelda.xsd" file.

djjuhasz replied:
Ah, okay — I'll update the comment. :)

jraddaoui commented on internal/sip/sip.go:
```diff
@@ -11,29 +11,49 @@ import (
-	if _, err := os.Stat(s.Path); err != nil {
-		return nil, fmt.Errorf("NewSIP: %v", err)
+	if _, err := os.Stat(path); err != nil {
+		return s, fmt.Errorf("NewSIP: %v", err)
```
We should tweak the error messages too, no longer called `NewSIP`.

jraddaoui reviewed (approved):
LGTM, just a couple nitpicks, thanks @djjuhasz!

jraddaoui reviewed (changes requested):
One thing I missed from https://github.com/artefactual-sdps/preprocessing-sfa/commit/6bd563e6f3cd237511b03ad2e9ddde05fe18c67a.

I think we should be validating `UpdatedAreldaMetadata.xml` for digitized AIPs and, with the changes in the commit above, we always validate `metadata.xml`. Now that you added  the `ManifestPath`, we should use that one calling the validation activity, here:

https://github.com/artefactual-sdps/preprocessing-sfa/blob/main/internal/workflow/preprocessing.go#L215

djjuhasz commented:
@jraddaoui I added a commit to fix the digitized AIP metadata validation — I think I'll keep it as a separate commit when I merge to main.

I also found another bug where the processing workflow wasn't halted when "verify manifest" failed.  I've added a third commit to fix this bug, which I'll squash into the first commit before merging.

jraddaoui commented:
Good catch, thanks @djjuhasz! I've done the squash and I'll merge to be able to deploy these changes.
